src/Microsoft.AspNet.Server.Kestrel/Http/Frame.cs

```diff
@@ -185,8 +196,91 @@ public bool HasResponseStarted
             get { return _responseStarted; }
         }
 
+        private void TimeoutRequest()
+        {
+            if (TransitionToState(FrameState.Timeout) == FrameState.Timeout)
+            {
+                SocketInput?.AbortAwaiting();
+            }
+        }
+        protected int TransitionToState(int state)
+        {
+#pragma warning disable CS0420 // A reference to a volatile field will not be treated as volatile
```

Comment:
Now you raise is... changed to normal `int` and changed all reads to `Volatile.Read` as while `volatile` gives compiler guarantees which is enough on Intel; it doesn't give CPU guarantees which may nor be enough on arm. (Don't know how the traditional C# memory memory model works on other chipssets). Also I'm using the value with `Interlocked` which wants `ref`

Also I've rolled two flags into one state and added some extras; so safety first.